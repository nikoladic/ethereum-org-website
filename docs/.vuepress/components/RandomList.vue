<template>
	<component :is="comp" :items="selected">
		<slot/>
	</component>
</template>

<script>
/*
*  Implements a component showing elements from the array `items` after shuffling it
*  using random.js APIs. The component can be used as a child component (e.g. RandomAppList.vue)
*  to implement a random selection of items.
*/
import DefaultList from './DefaultList.vue';
import {MersenneTwister19937, shuffle} from "random-js"

export default {
	props: {
		n: Number,
		items: Array,
		comp: {
			type: [String, Object],
			default: () => (DefaultList),
		}
	},

	data() {
		const random = MersenneTwister19937.autoSeed();
		return {random: random, selected: []}
	},

	created() {
		this.selectItems();
	},

	methods: {
		selectItems() {
			shuffle(this.random, this.items);
			this.selected = this.items.slice(0, this.n);
		}
	}
}
</script>
