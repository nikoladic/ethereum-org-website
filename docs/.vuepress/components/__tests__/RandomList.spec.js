import { shallowMount } from '@vue/test-utils'
import RandomList from "@/components/RandomList"
import { shuffle } from "random-js"

jest.mock('random-js');

describe('RandomList', () => {
  const items = [{
    name: "Bitcoin",
    description: "The 1st Crytocurrency in the world"
  }, {
    name: "Ethereum",
    description: "The smart contract blockchain"
  }, {
    name: "Litecoin",
    description: "A fork of Bitcoin"
  }, {
    name: "Steem",
    description: "The blockchain for social network"
  }]

  beforeAll(() => {
    shuffle.mockImplementation((seed, items) => {
      const first = items.shift()
      items.push(first)
      return items
    })
  })

  test('random display 2 out of 4 items', () => {
    let itemsCopy = [...items]
    const wrapper = shallowMount(RandomList, {
      propsData: {
        items: itemsCopy,
        n: 2
      }
    })
    expect(wrapper.vm.selected).toEqual(items.slice(1, 3)) // compare selected items
    expect(wrapper.element).toMatchSnapshot()
  })

  test('random display 3 out of 4 items', () => {
    let itemsCopy = [...items]
    const wrapper = shallowMount(RandomList, {
      propsData: {
        items: itemsCopy,
        n: 3
      }
    })
    expect(wrapper.vm.selected).toEqual(items.slice(1, 4)) // compare selected items
    expect(wrapper.element).toMatchSnapshot()
  })
})
