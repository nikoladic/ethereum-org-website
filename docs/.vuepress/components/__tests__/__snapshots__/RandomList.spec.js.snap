// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomList random display 2 out of 4 items 1`] = `
<default-list-stub
  items="[object Object],[object Object]"
/>
`;

exports[`RandomList random display 3 out of 4 items 1`] = `
<default-list-stub
  items="[object Object],[object Object],[object Object]"
/>
`;
